### DISCUSSION TITLE
Remark42 cluster version

### DISCUSSION THREAD

**Discussion by StarAurryon** [umputun/remark42#1346] [Ideas]
Hello there,

Just to announce that I have implemented a cluster backend for remark42 to use mongodb / s3 for image and engine store.

[https://github.com/Fenritec/remark42-cluster](https://github.com/Fenritec/remark42-cluster)

The implementation is complete for the image store and 80% for the engine store. So main functions are working and I need to write the tests to make it complete.

Feel free to give feedback.

The project is released under EUPL v1.2.

**Comment by umputun** [comment]
nice.

I have moved it to discussions as this is not really a bug report

**Comment by StarAurryon** [comment]
@umputun I have just one question. I saw that some query limit are hard-coded into the bolt engine backend. Shouldn't it be better to have standard behavior across all backends by having a common logic ?

**Comment by umputun** [comment]
I'm not sure what limits you mean exactly, but my guess would be 

```
const (
	// limits
	lastLimit = 1000
	userLimit = 500
)
```

Probably making it some struct and passing to the engine will be a good idea. Obviously, this way it can be also configured

**Reply by StarAurryon** [reply]
Yes that was that one.

**Reply by StarAurryon** [reply]
There is also this one in engine/bolt.go :
```
func (b *BoltDB) Create(comment store.Comment) (commentID string, err error) {
	bdb, err := b.db(comment.Locator.SiteID)
	if err != nil {
		return "", err
	}

	if b.checkFlag(FlagRequest{Locator: comment.Locator, Flag: ReadOnly}) {
		return "", fmt.Errorf("post %s is read-only", comment.Locator.URL)
	}
```
I can propose sth in a PR if you are ok

**Reply by umputun** [reply]
sure, prs are very welcome

**Reply by paskal** [reply]
I'm interested in your PR as well, and please let us know if in some place you would expect to see RPC storage support but didn't find it.

**Reply by StarAurryon** [reply]
What are you calling "RPC storage" ? RPC is already implemented for engine and image. Is it for the avatars?

**Reply by paskal** [reply]
You are right, all four of data, admin, image, and avatar storage support `rpc`, I've overlooked it looking at the documentation for some reason.

**Reply by paskal** [reply]
Actually after careful re-check I was wrong twice: first image storage has RPC storage but it was not mentioned in the parameters documentation (fixed in #1358), but second - avatar storage missing RPC store support and example implementation. I propose to create an issue about it if you'll need it.

**Reply by StarAurryon** [reply]
For our project it is not needed as the avatar lib you rely on has a mongo storage backend. But we can definitely make a RPC backend in the future.
